sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/BatchSpannerRead.java

```diff
@@ -73,18 +74,16 @@ public static BatchSpannerRead create(
         .apply(
             "Generate Partitions",
             ParDo.of(new GeneratePartitionsFn(getSpannerConfig(), txView)).withSideInputs(txView))
-        .apply("Shuffle partitions", Reshuffle.<Partition>viaRandomKey())
         .apply(
             "Read from Partitions",
             ParDo.of(new ReadFromPartitionFn(getSpannerConfig(), txView)).withSideInputs(txView));
   }
 
   @VisibleForTesting
-  static class GeneratePartitionsFn extends DoFn<ReadOperation, Partition> {
+  static class GeneratePartitionsFn extends DoFn<ReadOperation, List<Partition>> {
```

Comment:
Hello @boyuanzz I have been working in trying to get this part done but I have encountered a problem with the approach. I have changed the code to a single DoFn to remove `GeneratePartitionsFn` as you mentioned, now my DoFn looks like this `private static class ReadFromPartitionFn extends DoFn<ReadOperation, Struct>`. But I'm having problems getting `@GetInitialRestriction `and `@SplitRestriction` right, in order to know the size of the Partition List we are going to process I need to use the side input `c.sideInput(txView);` to process the ReadOperation, is there a way to get the SideInput in `@GetInitialRestriction` or `@SplitRestriction` to get the List Size and the amount of partitions we are going to process? or the approach you were suggesting was something different? Thanks in advance.

Reply:
Sorry for the late. 
`Splittable DoFn` supports sideinput just like normal `DoFn`. So you can access sideinput from `@GetInitialRestriction` and `@SplitRestriction` just like from `@ProcessElement`.

Reply:
Hi @boyuanzz , thanks for your response, I have been trying out that option, trying to access the sideinput either from `@GetInitialRestriction` or `@SplitRestriction`, both using `@SideInput` and` ProcessContext,` but in both cases I get an error specifying that `Illegal parameter type: ProcessContextParameter{}` or `Illegal parameter type: SideInputParameter`, checking the code for `DoFnSignatures` I can see `@SideInput` is not in the `ALLOWED_SPLIT_RESTRICTION_PARAMETERS` or `ALLOWED_GET_INITIAL_RESTRICTION_PARAMETERS`. I don’t know if I maybe missing something in how to use the side parameters?

Reply:
Let me double check and I'll back to here when I get an answer.

Reply:
btw, can you share the whole stacktrace for the error you get?

Reply:
java.lang.IllegalArgumentException: org.apache.beam.sdk.io.gcp.spanner.BatchSpannerRead$ReadFromPartitionFn, @SplitRestriction splitRestriction(OffsetRange, PCollectionView, OutputReceiver): Illegal parameter type: SideInputParameter{elementT=org.apache.beam.sdk.values.PCollectionView<org.apache.beam.sdk.io.gcp.spanner.Transaction>, sideInputId=tx}
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures$ErrorReporter.throwIllegalArgument(DoFnSignatures.java:2399)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.checkParameterOneOf(DoFnSignatures.java:1292)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.analyzeSplitRestrictionMethod(DoFnSignatures.java:1841)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.parseSignature(DoFnSignatures.java:722)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.lambda$getSignature$0(DoFnSignatures.java:282)
	at java.util.HashMap.computeIfAbsent(HashMap.java:1127)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.getSignature(DoFnSignatures.java:282)
	at org.apache.beam.sdk.transforms.ParDo.validate(ParDo.java:613)
	at org.apache.beam.sdk.transforms.ParDo.of(ParDo.java:402)
	at org.apache.beam.sdk.io.gcp.spanner.BatchSpannerRead.expand(BatchSpannerRead.java:79)
	at org.apache.beam.sdk.io.gcp.spanner.BatchSpannerRead.expand(BatchSpannerRead.java:43)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:548)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:499)
	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:370)
	at org.apache.beam.sdk.io.gcp.spanner.SpannerIO$ReadAll.expand(SpannerIO.java:518)
	at org.apache.beam.sdk.io.gcp.spanner.SpannerIO$ReadAll.expand(SpannerIO.java:395)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:548)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:499)
	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:370)
	at org.apache.beam.sdk.io.gcp.spanner.SpannerIOReadTest.readAllPipeline(SpannerIOReadTest.java:305)

Reply:
Hi @boyuanzz 
Retaking the idea in this issue, I looked for an example of using side inputs in the SplitRestriction and GetInitialRestriction methods but I haven’t find any. In case that it's not possible to use the sideinput there to estimate the amount of data I think the best way to get a good estimation of the amount of data to split would be to keep the logic in two doFn, as we get the Partitions and the size of the list of partitions that we can work with.

Reply:
I have created https://issues.apache.org/jira/browse/BEAM-12522 for tracking the sideinput with Splittable DoFn issue. I'm trying to work on a quick fix, though I'm not sure whether the fix would be simple. In order to unblock you, you can put a TODO(BEAM-12522) over the expansion saying that we can do so when the sideinput is supported.

Reply:
Added TODO message to GeneratePartitionsFn mentioning it should be integrated into ReadFromPartitionFn when the issue is solved